params/governance_paramset.go

```diff
@@ -0,0 +1,419 @@
+	items map[int]interface{}
+}
+
+func NewGovParamSet() *GovParamSet {
+	return &GovParamSet{
+		items: make(map[int]interface{}),
+	}
+}
+
+func NewGovParamSetStrMap(items map[string]interface{}) (*GovParamSet, error) {
+	p := NewGovParamSet()
+
+	for name, value := range items {
+		key, ok := govParamNames[name]
+		if !ok {
+			return nil, fmt.Errorf("Unknown governance param '%s'", name)
+		}
+		err := p.set(key, value)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return p, nil
+}
+
+func NewGovParamSetIntMap(items map[int]interface{}) (*GovParamSet, error) {
+	p := NewGovParamSet()
+
+	for key, value := range items {
+		err := p.set(key, value)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return p, nil
+}
+
+func NewGovParamSetChainConfig(config *ChainConfig) (*GovParamSet, error) {
+	items := make(map[int]interface{})
+	if config.Istanbul != nil {
+		items[Epoch] = config.Istanbul.Epoch
+		items[Policy] = config.Istanbul.ProposerPolicy
+		items[CommitteeSize] = config.Istanbul.SubGroupSize
+	}
+	items[UnitPrice] = config.UnitPrice
+	if config.Governance != nil {
+		items[GoverningNode] = config.Governance.GoverningNode
+		items[GovernanceMode] = config.Governance.GovernanceMode
+		if config.Governance.Reward != nil {
+			if config.Governance.Reward.MintingAmount != nil {
+				items[MintingAmount] = config.Governance.Reward.MintingAmount.String()
+			}
+			items[Ratio] = config.Governance.Reward.Ratio
+			items[UseGiniCoeff] = config.Governance.Reward.UseGiniCoeff
+			items[DeferredTxFee] = config.Governance.Reward.DeferredTxFee
+			items[StakeUpdateInterval] = config.Governance.Reward.StakingUpdateInterval
+			items[ProposerRefreshInterval] = config.Governance.Reward.ProposerUpdateInterval
+			if config.Governance.Reward.MinimumStake != nil {
+				items[MinimumStake] = config.Governance.Reward.MinimumStake.String()
+			}
+		}
+	}
+
+	return NewGovParamSetIntMap(items)
+}
+
+func (p *GovParamSet) set(key int, value interface{}) error {
+	ty, ok := govParamTypes[key]
+	if !ok {
+		return errors.New("Unknown governance param key")
+	}
+	parsed, ok := ty.ParseValue(value)
+	if !ok {
+		return errors.New("Malformed governance param value")
+	}
+	p.items[key] = parsed
+	return nil
+}
+
+func (p *GovParamSet) StrMap() map[string]interface{} {
+	m := map[string]interface{}{}
+	for key, value := range p.items {
+		m[govParamNamesReverse[key]] = value
+	}
+	return m
+}
+
+func (p *GovParamSet) IntMap() map[int]interface{} {
+	m := map[int]interface{}{}
+	for key, value := range p.items {
+		m[key] = value
+	}
+	return m
+}
+
+func (p *GovParamSet) Get(key int) (interface{}, bool) {
+	v, ok := p.items[key]
+	return v, ok
+}
+
+// If the value is not set, logs a warning and returns the default value.
+func (p *GovParamSet) MustGet(key int) interface{} {
+	if v, ok := p.items[key]; ok {
+		return v
+	}
+	if d, ok := govParamDefaults[key]; ok {
+		// If key is a known parameter, fallback to default
+		return d
+	}
```

Comment:
The comment says `if the value is not set, logs a warning`. According to the comments, shouldn't a warning log be added? 

By the way, if there is no value corresponding to the `key`, it seems it reads the default value. If the default value changes, it gets a different value compared to the past. Is it okay to read different value?

Reply:
I mean, if the codebase's default value is changes, it reads the difference value compared to the past. Are you meaning that the constants will not be changed?

Reply:
Understood. The fallback value is there to make sure calls like `goverance.Epoch()` does not panic. These nominal getters are widely used throughout our codebase. The problem is when there is no value.

We have some choices here.

1. Do not provide MustGet and nominal getters at all
    - I'd say it's the safest way. All users are aware that some governance parameter can be undefined and do their own error handling.
    - Proper error handling is required. 
    ```go
    var epoch uint64
    if val, ok := gov.Params().Get(params.Epoch); ok {
        epoch = val.(uint64)
    } else {
        logger.Crit("Cannot get istanbul.epoch")
    }
    ```
2. MustGet returns an empty value like `0` or `""`
    - Empty value may be silently used. Potentially hard-to-catch bug.
    ```go
    epoch := gov.Params().Epoch() // could be 0 on error
    ```
3. MustGet returns designated default value (i.e. params.DefaultXX)
    - It is hard to distinguish between fallback value and actual value.
    - Default value may be silently used.
    - The default may change later. If some code might be relying on the default value (even if the developer didn't know). Potentially hard-to-catch bug.
    ```go
    epoch := gov.Params().Epoch() // could be 86400 on error
    ```
4. Each GovParamSet instance holds an explicit default value
    - For instance, ChainConfig can be used as default value.
    - This behavior is similar to current default Governance, because currentSet is first populated with chainConfig and then overwritten by DB content.
    ```go
    // governance/engine.go
    func NewMyEngine(config *params.ChainConfig, ...) *MyEngine {
        return &MyEngine{
            ChainConfig: config,
            DefaultSet: GovParamSetChainConfig(config),
        }
    }
    func (gov *Engine) Params() *params.GovParamSet {
        data := gov.ReadGovernance()
        return GovParamSetStrMap(data, gov.DefaultSet)
    }
    // params/governance_paramset.go
    func (p *GovParamSet) MustGet(key int) interface{} {
        if v, ok := p.Get(key); ok {
            return v
        } else if v, ok := p.Defaults.Get(key); ok {
            return v
        } else {
            logger.Crit("no default value for", key)
        }
    }            
    ```
    ```go
    epoch := gov.Params().Epoch() // could be config.Istanbul.Epoch on error
    ```

Current PR is choice3. But I think moving to choice1 or choice4 would be safer.
@yoomee1313 What do you think?

Reply:
Thanks for organizing it so well. I agree with you that choice 1 or choice 4 is safer. Will you fix the code as choice 1 or choice4?

Reply:
Yes, preferrably in this PR.

Reply:
@yoomee1313 How about going for choice1 in this PR (leave only `Get()`), then add MustGet and nominal getters later? If you are okay, I will add commits.

Reply:
@blukat29 Sure, I'm okay with option1.